// @jridgewell/source-map/dist/source-map.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var traceMapping = require('@jridgewell/trace-mapping');
var genMapping = require('@jridgewell/gen-mapping');

class SourceMapConsumer {
    constructor(map, mapUrl) {
        const trace = (this._map = new traceMapping.AnyMap(map, mapUrl));
        this.file = trace.file;
        this.names = trace.names;
        this.sourceRoot = trace.sourceRoot;
        this.sources = trace.resolvedSources;
        this.sourcesContent = trace.sourcesContent;
        this.version = trace.version;
    }
    static fromSourceMap(map, mapUrl) {
        // This is more performant if we receive
        // a @jridgewell/source-map SourceMapGenerator
        if (map.toDecodedMap) {
            return new SourceMapConsumer(map.toDecodedMap(), mapUrl);
        }
        // This is a fallback for `source-map` and `source-map-js`
        return new SourceMapConsumer(map.toJSON(), mapUrl);
    }
    get mappings() {
        return traceMapping.encodedMappings(this._map);
    }
    originalPositionFor(needle) {
        return traceMapping.originalPositionFor(this._map, needle);
    }
    generatedPositionFor(originalPosition) {
        return traceMapping.generatedPositionFor(this._map, originalPosition);
    }
